virtual_ecosystem/models/soil/microbial_groups.py

```diff
@@ -99,14 +106,112 @@ class MicrobialGroupConstants:
+    ):
+        """Class method to build the microbial group including enzyme information.
+
+        Args:
+            group_config: The config details for microbial group in question.
+            enzyme_classes: Details of the enzyme classes used by the soil model.
+        """
+
+        return cls(
+            **group_config,
+            synthesis_nutrient_ratios=calculate_new_biomass_average_nutrient_ratios(
+                name=group_config["name"],
+                c_n_ratio=group_config["c_n_ratio"],
+                c_p_ratio=group_config["c_p_ratio"],
+                enzyme_production=group_config["enzyme_production"],
+                enzyme_classes=enzyme_classes,
+            ),
+        )
+
+    def find_enzyme_substrates(self) -> list[str]:
+        """Substrates that the microbial group produces enzymes for."""
+
+        return [
+            substrate
+            for substrate, production in self.enzyme_production.items()
+            if production > 0.0
+        ]
+
+
+def calculate_new_biomass_average_nutrient_ratios(
+    name: str,
+    c_n_ratio: float,
+    c_p_ratio: float,
+    enzyme_production: dict[str, float],
+    enzyme_classes: dict[str, EnzymeConstants],
+) -> dict[str, float]:
+    """Calculate average carbon nutrient ratios of the newly synthesised biomass.
+
+    Microbes have to synthesise cellular biomass as well as extracellular enzymes. This
+    method calculates average nutrient ratio of this total biomass synthesis by
+    calculating the average weighted by the relative production allocation to each
+    enzyme class and cellular growth.
+
+    Args:
+        name: Name of the microbial group.
+        c_n_ratio: Ratio of carbon to nitrogen for the microbial group's cellular
+            biomass.
+        c_p_ratio: Ratio of carbon to nitrogen for the microbial group's cellular
+            biomass.
+        enzyme_production: Details of the enzymes produced by the microbial group, i.e.
+            which substrates are enzymes produced for, and how much (relative to
+            cellular synthesis)
+        enzyme_classes: Details of the enzyme classes used by the soil model.
+    """
+
+    total_enzyme_allocation = sum(enzyme_production.values())
+
+    enzyme_c_n_weighted = sum(
+        enzyme_classes[f"{name}_{substrate}"].c_n_ratio
+        * allocation
+        / total_enzyme_allocation
+        for substrate, allocation in enzyme_production.items()
+    )
+
+    enzyme_c_p_weighted = sum(
+        enzyme_classes[f"{name}_{substrate}"].c_p_ratio
+        * allocation
+        / total_enzyme_allocation
+        for substrate, allocation in enzyme_production.items()
+    )
+
+    return {
+        "nitrogen": (c_n_ratio + enzyme_c_n_weighted * total_enzyme_allocation)
+        / (1.0 + total_enzyme_allocation),
+        "phosphorus": (c_p_ratio + enzyme_c_p_weighted * total_enzyme_allocation)
+        / (1.0 + total_enzyme_allocation),
+    }
```

Comment:
Please, let someone check my maths, but I think that you can take the dividing `total_enzyme_allocation` out of the sums, in which case it will cancel with the multiplying one  below:
```suggestion
    enzyme_c_n_weighted = sum(
        enzyme_classes[f"{name}_{substrate}"].c_n_ratio
        * allocation
        for substrate, allocation in enzyme_production.items()
    )

    enzyme_c_p_weighted = sum(
        enzyme_classes[f"{name}_{substrate}"].c_p_ratio
        * allocation
        for substrate, allocation in enzyme_production.items()
    )

    return {
        "nitrogen": (c_n_ratio + enzyme_c_n_weighted)
        / (1.0 + total_enzyme_allocation),
        "phosphorus": (c_p_ratio + enzyme_c_p_weighted)
        / (1.0 + total_enzyme_allocation),
    }
```

Reply:
Yes, you are right! I think I was overly focused on the line-by-line logic so I lost the big picture